tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -7,125 +7,438 @@
-            where TCarrierGetter : struct, ICarrierGetter<TCarrier>
+        internal static string CreateTraceStateHeader(SpanContext context)
         {
-            spanContext = null;
+            var samplingPriority = SamplingPriorityToString(context.TraceContext?.SamplingPriority ?? context.SamplingPriority);
+            StringBuilder? sb = null;
 
-            var traceParent = ParseUtility.ParseString(carrier, carrierGetter, TraceParent)?.Trim();
-            if (!string.IsNullOrEmpty(traceParent))
+            try
             {
-                // We found a trace parent (we are reading from the Http Headers)
+                sb = StringBuilderCache.Acquire(100);
+                sb.Append("dd=");
 
-                /* (https://www.w3.org/TR/trace-context/)
-                    Valid traceparent when caller sampled this request:
+                if (samplingPriority != null)
+                {
+                    sb.Append("s:").Append(samplingPriority).Append(';');
+                }
 
-                    Value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-01
-                    base16(version) = 00
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 01  // sampled
-                    Valid traceparent when caller didn’t sample this request:
+                if (!string.IsNullOrWhiteSpace(context.Origin))
+                {
+                    sb.Append("o:").Append(context.Origin).Append(';');
+                }
+
+                if (context.TraceContext?.Tags?.ToArray() is { Length: > 0 } tags)
+                {
+                    foreach (var tag in tags)
+                    {
+                        if (tag.Key.StartsWith("_dd.p.", StringComparison.Ordinal))
+                        {
+#if NETCOREAPP
+                            var key = tag.Key.AsSpan(start: 6);
+#else
+                            var key = tag.Key.Substring(startIndex: 6);
+#endif
 
-                    Value = 00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-00
-                    base16(version) = 00
-                    base16(trace-id) = 4bf92f3577b34da6a3ce929d0e0e4736
-                    base16(parent-id) = 00f067aa0ba902b7
-                    base16(trace-flags) = 00  // not sampled
-                 */
+                            sb.Append("t.").Append(key).Append(':').Append(tag.Value).Append(';');
+                        }
+                    }
+                }
 
-                if (traceParent!.Length != 55 || traceParent[2] != '-' || traceParent[35] != '-' || traceParent[52] != '-')
+                if (sb.Length == 3)
                 {
-                    return false;
+                    // "dd=", we never appended anything
+                    return string.Empty;
```

Comment:
```suggestion
                    StringBuilderCache.Release(sb);
                    return string.Empty;
```

Reply:
Just realized you already have a try/finally. If you think an exception may be thrown here you could keep it, but maybe you can just call `StringBuilderCache.Release` on every return

Reply:
I was going with the "just make sure it always happens to be safe" technical. No strong feeling about it, though, if you prefer I change it.

Reply:
Or not? Since the reference is saved in a thread local...